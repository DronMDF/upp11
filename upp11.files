LICENSE
README.md
upp11.h
Makefile
test/testupp.cpp
test/testfailures.cpp
test/testuppdetail.cpp
