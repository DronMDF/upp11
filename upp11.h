
#pragma once
#include <list>
#include <iostream>

namespace upp11 {

class TestInvoker {
public:
	virtual void invoke() = 0;
};

class TestCollection {
private:
	std::list<TestInvoker *> tests;

	static TestCollection &getInstance() {
		static TestCollection collection;
		return collection;
	};

public:
	static void addTest(class TestInvoker *test)
	{
		TestCollection &collection = getInstance();
		collection.tests.push_back(test);
	}

	static void runAllTests()
	{
		TestCollection &collection = getInstance();
		for (auto *t: collection.tests) {
			t->invoke();
		}
	}
};

template <typename T>
class TestInvokerImpl : public TestInvoker {
public:
	TestInvokerImpl() {
		TestCollection::addTest(this);
	}

private:
	void invoke() override {
		T test;
		test();
	}
};

} // end of namespace upp11

#define UP_RUN() \
	upp11::TestCollection::runAllTests()

#define UP_TEST(name) \
class Test##name { \
public: \
	void operator()(); \
}; \
static upp11::TestInvokerImpl<Test##name> test##name##invoker; \
void Test##name::operator()()

#define UP_FAIL(msg) \
	std::cout << msg << std::endl;
