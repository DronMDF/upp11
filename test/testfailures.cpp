#include <upp11.h>

using namespace std;

UP_SUITE_BEGIN(suiteCheckpoints)

UP_TEST(UnhandledExceptionInTestShouldCheckpointed)
{
	throw false;
}

struct throw_fixture {
	virtual ~throw_fixture() = default;
	throw_fixture() {
		throw runtime_error("setUp exception for checkpoint");
	}
};

UP_FIXTURE_TEST(ExceptionInFixtureShouldCheckpointed, throw_fixture)
{
}

struct zero_fixture {
	virtual ~zero_fixture() = default;
	int divide(int a, int b) const;
};

UP_FIXTURE_TEST(DivizionByZeroInTestShouldCheckpointed, zero_fixture)
{
	divide(10, 0);
}

int zero_fixture::divide(int a, int b) const {
	return a / b;
}

struct deref_fixture {
	virtual ~deref_fixture() = default;
	int deref(int *a) const;
};

UP_FIXTURE_TEST(SegFaultInTestShouldCheckpointed, deref_fixture)
{
	deref(nullptr);
}

int deref_fixture::deref(int *a) const {
	return *a;
}

UP_TEST(AssertIsCheckpoint)
{
	UP_ASSERT(true);
	throw false;
}

UP_TEST(AssertEqualIsCheckpoint)
{
	UP_ASSERT_EQUAL(1, 1);
	throw false;
}

UP_TEST(AssertNeIsCheckpoint)
{
	UP_ASSERT_NE(1, 0);
	throw false;
}

UP_TEST(AssertExceptionIsCheckpoint)
{
	UP_ASSERT_EXCEPTION(runtime_error, []{ throw runtime_error("true"); });
	throw false;
}

UP_TEST(AssertExceptionMessageIsCheckpoint)
{
	UP_ASSERT_EXCEPTION(runtime_error, "true", []{ throw runtime_error("true"); });
	throw false;
}

struct throw_param_fixture {
	virtual ~throw_param_fixture() = default;
	bool param() const {
		throw false;
	}
};

UP_FIXTURE_TEST(CheckpointShoildBeFixedBeforeArgumentsEvaluate, throw_param_fixture)
{
	UP_ASSERT(param());
}

UP_TEST(ExplicitCheckpointShouldBe)
{
	UP_CHECKPOINT("user checkpoint");
	throw false;
}

UP_FIXTURE_TEST(SecondSignalShouldBeCatched, deref_fixture)
{
	UP_ASSERT_EQUAL(deref(nullptr), 42);
}

UP_SUITE_END()

UP_SUITE_BEGIN(suiteAssertEqual)

UP_TEST(ShouldFailByNoEqual)
{
	UP_ASSERT_EQUAL(1, 0);
}

UP_SUITE_END()

UP_SUITE_BEGIN(suiteAssertNe)

UP_TEST(ShouldFailByEqual)
{
	UP_ASSERT_NE(1, 1);
}

UP_SUITE_END()

UP_SUITE_BEGIN(suiteAssertExceptionWithMessage)

UP_TEST(ShouldFailByType)
{
	UP_ASSERT_EXCEPTION(int, "hello", []{
		throw runtime_error("goodby");
	});
}

UP_TEST(ShouldFailByNoThrow)
{
	UP_ASSERT_EXCEPTION(runtime_error, "hello", []{
		// no throw
	});
}

UP_TEST(ShouldFailByNotEqualMessage)
{
	UP_ASSERT_EXCEPTION(exception, "message", []{
		throw runtime_error("another message");
	});
}

UP_TEST(ShouldFailByChildException)
{
	UP_ASSERT_EXCEPTION(overflow_error, "message", []{
		throw runtime_error("message");
	});
}

UP_SUITE_END()

UP_SUITE_BEGIN(suiteAssertException)

UP_TEST(ShouldFailByType)
{
	UP_ASSERT_EXCEPTION(int, []{
		throw runtime_error("goodby");
	});
}

UP_TEST(ShouldFailByNoThrow)
{
	UP_ASSERT_EXCEPTION(runtime_error, []{
		// no throw
	});
}

UP_SUITE_END()

UP_MAIN()
