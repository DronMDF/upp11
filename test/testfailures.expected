test/testfailures.cpp(110): check equal (1, 0) failed
	1 vs 0
suiteAssertEqual::ShouldFailByNoEqual: FAIL
test/testfailures.cpp(169): expected exception runtime_error not throw
suiteAssertException::ShouldFailByNoThrow: FAIL
test/testfailures.cpp(162): expected exception int not throw
suiteAssertException::ShouldFailByType: FAIL
test/testfailures.cpp(151): expected exception overflow_error("message") not throw
suiteAssertExceptionWithMessage::ShouldFailByChildException: FAIL
test/testfailures.cpp(137): expected exception runtime_error("hello") not throw
suiteAssertExceptionWithMessage::ShouldFailByNoThrow: FAIL
test/testfailures.cpp(144): check exception exception("message") failed
	catched exception: "another message"
suiteAssertExceptionWithMessage::ShouldFailByNotEqualMessage: FAIL
test/testfailures.cpp(130): expected exception int is not child of std::exception
suiteAssertExceptionWithMessage::ShouldFailByType: FAIL
test/testfailures.cpp(119): check not equal (1, 1) failed
	1 vs 1
suiteAssertNe::ShouldFailByEqual: FAIL
unexpected test termination
test/testfailures.cpp(59): last checkpoint: UP_ASSERT_EQUAL
suiteCheckpoints::AssertEqualIsCheckpoint: FAIL
unexpected test termination
test/testfailures.cpp(71): last checkpoint: UP_ASSERT_EXCEPTION
suiteCheckpoints::AssertExceptionIsCheckpoint: FAIL
unexpected test termination
test/testfailures.cpp(77): last checkpoint: UP_ASSERT_EXCEPTION
suiteCheckpoints::AssertExceptionMessageIsCheckpoint: FAIL
unexpected test termination
test/testfailures.cpp(53): last checkpoint: UP_ASSERT
suiteCheckpoints::AssertIsCheckpoint: FAIL
unexpected test termination
test/testfailures.cpp(65): last checkpoint: UP_ASSERT_NE
suiteCheckpoints::AssertNeIsCheckpoint: FAIL
unexpected test termination
test/testfailures.cpp(90): last checkpoint: UP_ASSERT
suiteCheckpoints::CheckpointShoildBeFixedBeforeArgumentsEvaluate: FAIL
unexpected test termination: Test terminated by signal
test/testfailures.cpp(28): last checkpoint: run test
suiteCheckpoints::DivizionByZeroInTestShouldCheckpointed: FAIL
unexpected test termination: setUp exception for checkpoint
test/testfailures.cpp(19): last checkpoint: fixture setUp
suiteCheckpoints::ExceptionInFixtureShouldCheckpointed: FAIL
unexpected test termination
test/testfailures.cpp(95): last checkpoint: user checkpoint
suiteCheckpoints::ExplicitCheckpointShouldBe: FAIL
unexpected test termination: Test terminated by signal
test/testfailures.cpp(101): last checkpoint: UP_ASSERT_EQUAL
suiteCheckpoints::SecondSignalShouldBeCatched: FAIL
unexpected test termination: Test terminated by signal
test/testfailures.cpp(42): last checkpoint: run test
suiteCheckpoints::SegFaultInTestShouldCheckpointed: FAIL
unexpected test termination
test/testfailures.cpp(7): last checkpoint: run test
suiteCheckpoints::UnhandledExceptionInTestShouldCheckpointed: FAIL
Run 20 tests with 20 failures
